wire-runtime/src/main/java/com/squareup/wire/TypeAdapter.java

```diff
@@ -0,0 +1,369 @@
+/*
+ * Copyright 2015 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire;
+
+import java.io.IOException;
+import okio.ByteString;
+
+import static java.lang.Double.doubleToLongBits;
+import static java.lang.Float.floatToIntBits;
+
+public abstract class TypeAdapter<E> {
+  static final int TYPE_VARINT = 0;
+  static final int TYPE_FIXED64 = 1;
+  static final int TYPE_LEN_DELIMITED = 2;
+  static final int TYPE_START_GROUP = 3;
+  static final int TYPE_END_GROUP = 4;
+  static final int TYPE_FIXED32 = 5;
+
+  static final int FIXED_BOOL_SIZE = 1;
+  static final int FIXED_32_SIZE = 4;
+  static final int FIXED_64_SIZE = 8;
+
+  static int utf8Length(String s) {
+    int count = 0;
+    for (int i = 0, length = s.length(); i < length; i++) {
+      char ch = s.charAt(i);
+      if (ch <= 0x7F) {
+        count++;
+      } else if (ch <= 0x7FF) {
+        count += 2;
+      } else if (Character.isHighSurrogate(ch)) {
+        count += 4;
+        ++i;
```

Comment:
There are badly-formed strings that will cause problems here. I'm not sure what a proto serializer is supposed to do about such strings, though.